README.md
# Personal Bugdet Application

## Objective
- Get a handle on one's personal finances.
- Take in expense information.
- Save previous expenses.
- Analyze all previous data to identify trends in expenses.
- Take advantage of interoperating with personal financial institutions.
- Written in Rust.
- Expose a web interface to the internet.
- Self-hosted.

## Features
- [ ] Progressive Web App.
- [ ] Take in expense information.
- [ ] Save previous expenses.
- [ ] Analyze all previous data to identify trends in expenses.
- [ ] Take advantage of interoperating with personal financial institutions.
- [ ] Written in Rust.
- [ ] No JavaScript.
- [ ] Self-hosted.
- [ ] Expose a web interface to the internet.


## Installation
- Download the executable.
- Run the executable.
- Access the application at `localhost:8123`.
- Access the application from the internet at `TBD`.

## Usage
- Login or create an account.
- Enter expense information.
- View previous expenses.
- View trends in expenses.
- Link a personal financial institution.

## Contributing
- Fork the repository.
- Make changes.
- Submit a pull request.

## Initial Design

![Initial Design](./assets/docs/init_design.png)

	




CHANGELOG.md

# CHANGELOG										       
											       
Author: Hunter, Christerpher								       
											       
All notable changes will be appended here.						       
											       
This project, henceforth, will recongnize [semantic versioning](https://semver.org/).	       
											       
## [⭐.✴️.✳️] - YEAR MONTH DAY								       
											       
Here we write upgrade and change notes.						       
											       
⭐ MAJOR version when you make incompatible API changes,				   
											       
✴️ MINOR version when you add functionality in a backwards compatible manner		       
											       
✳️ PATCH version when you make backwards compatible bug fixes.				       
											       
--------------------------------------

✳️[0.0.1] - 2023 JUNE 25

- Automatic initialization of the project structure 



tests/integration_test.rs
use actix_web::{http, test, web, App};
use personal_budget_app_rs::{
    About, Budget, Debt, HelloTemplate, Income, Investment, NotFound, Retirement, Savings,
};

#[actix_web::test]
async fn test_index() {
    let app = test::init_service(App::new().route(
        "/",
        web::get().to(|| async {
            HelloTemplate {
                title: "Hunter",
                index: "Home",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_about() {
    let app = test::init_service(App::new().route(
        "/about",
        web::get().to(|| async {
            About {
                title: "About",
                about: "About",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/about").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_debt() {
    let app = test::init_service(App::new().route(
        "/debt",
        web::get().to(|| async {
            Debt {
                title: "Debt",
                debt: "Debt",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/debt").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_finances() {
    let app = test::init_service(App::new().route(
        "/finances",
        web::get().to(|| async {
            Savings {
                title: "Savings",
                savings: "Savings",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/finances").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_income() {
    let app = test::init_service(App::new().route(
        "/income",
        web::get().to(|| async {
            Income {
                title: "Income",
                income: "Income",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/income").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_investment() {
    let app = test::init_service(App::new().route(
        "/investment",
        web::get().to(|| async {
            Investment {
                title: "Investment",
                investment: "Investment",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/investment").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_retirement() {
    let app = test::init_service(App::new().route(
        "/retirement",
        web::get().to(|| async {
            Retirement {
                title: "Retirement",
                retirement: "Retirement",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/retirement").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_budget() {
    let app = test::init_service(App::new().route(
        "/budget",
        web::get().to(|| async {
            Budget {
                title: "Budget",
                budget: "Budget",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/budget").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_404() {
    let app = test::init_service(App::new().route(
        "/404",
        web::get().to(|| async {
            NotFound {
                title: "404",
                not_found: "404",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/404").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(resp.status(), http::StatusCode::OK);
}

#[actix_web::test]
async fn test_index_returns_html() {
    let app = test::init_service(App::new().route(
        "/",
        web::get().to(|| async {
            HelloTemplate {
                title: "Hunter",
                index: "Home",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_about_returns_html() {
    let app = test::init_service(App::new().route(
        "/about",
        web::get().to(|| async {
            About {
                title: "About",
                about: "About",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/about").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_debt_returns_html() {
    let app = test::init_service(App::new().route(
        "/debt",
        web::get().to(|| async {
            Debt {
                title: "Debt",
                debt: "Debt",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/debt").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_finances_returns_html() {
    let app = test::init_service(App::new().route(
        "/finances",
        web::get().to(|| async {
            Savings {
                title: "Savings",
                savings: "Savings",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/finances").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_income_returns_html() {
    let app = test::init_service(App::new().route(
        "/income",
        web::get().to(|| async {
            Income {
                title: "Income",
                income: "Income",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/income").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_retirement_returns_html() {
    let app = test::init_service(App::new().route(
        "/retirement",
        web::get().to(|| async {
            Retirement {
                title: "Retirement",
                retirement: "Retirement",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/retirement").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_investment_returns_html() {
    let app = test::init_service(App::new().route(
        "/investment",
        web::get().to(|| async {
            Investment {
                title: "Investment",
                investment: "Investment",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/investment").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_budget_returns_html() {
    let app = test::init_service(App::new().route(
        "/budget",
        web::get().to(|| async {
            Budget {
                title: "Budget",
                budget: "Budget",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/budget").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}

#[actix_web::test]
async fn test_404_returns_html() {
    let app = test::init_service(App::new().route(
        "/404",
        web::get().to(|| async {
            NotFound {
                title: "404",
                not_found: "404",
            }
        }),
    ))
    .await;

    let req = test::TestRequest::get().uri("/404").to_request();
    let resp = test::call_service(&app, req).await;

    assert_eq!(
        resp.headers().get(http::header::CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
}


src/lib.rs
use actix_web::{get, http::StatusCode, HttpResponse};
use askama::Template;

#[get("/home_button")]
pub async fn home_button() -> HttpResponse {
    println!("Home button clicked");

    let response: &str = "
    <button style=\"background-color: #4CAF50; /* Green */
  border: none;
  border-radius: 8px;
  color: white;
  padding: 15px 32px;
  text-align: center;
  font-size: 16px;
  margin: 4px 2px;
  cursor: pointer;\"
        >Home</button>
    ";
    HttpResponse::build(StatusCode::OK)
        .content_type("text/html; charset=utf-8")
        .body(response.to_string())
}

#[derive(Template)]
#[template(path = "index.html")]
pub struct HelloTemplate<'a> {
    pub title: &'a str,
    pub index: &'a str,
}

#[derive(Template)]
#[template(path = "about.html")]
pub struct About<'a> {
    pub title: &'a str,
    pub about: &'a str,
}

#[derive(Template)]
#[template(path = "debt.html")]
pub struct Debt<'a> {
    pub title: &'a str,
    pub debt: &'a str,
}

#[derive(Template)]
#[template(path = "savings.html")]
pub struct Savings<'a> {
    pub title: &'a str,
    pub savings: &'a str,
}

#[derive(Template)]
#[template(path = "income.html")]
pub struct Income<'a> {
    pub title: &'a str,
    pub income: &'a str,
}

#[derive(Template)]
#[template(path = "retirement.html")]
pub struct Retirement<'a> {
    pub title: &'a str,
    pub retirement: &'a str,
}

#[derive(Template)]
#[template(path = "investment.html")]
pub struct Investment<'a> {
    pub title: &'a str,
    pub investment: &'a str,
}

#[derive(Template)]
#[template(path = "budget.html")]
pub struct Budget<'a> {
    pub title: &'a str,
    pub budget: &'a str,
}

#[derive(Template)]
#[template(path = "404.html")]
pub struct NotFound<'a> {
    pub title: &'a str,
    pub not_found: &'a str,
}


src/main.rs
mod logic;

use actix_web::{web, App, HttpServer};

use personal_budget_app_rs::{
    home_button, About, Budget, Debt, HelloTemplate, Income, Investment, NotFound, Retirement,
    Savings,
};

use crate::logic::savings::savings_post;

use simplelog::{CombinedLogger, TermLogger, WriteLogger};
use std::fs::File;

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let titled: &str = "Home";
    let abouted: &str = "About";
    let debted: &str = "Debt";
    let savinged: &str = "Savings";
    let incomed: &str = "Income";
    let invested: &str = "Investment";
    let retired: &str = "Retirement";
    let budgeted: &str = "Budget";
    let not_founded: &str = "404";

    CombinedLogger::init(vec![
        TermLogger::new(
            log::LevelFilter::Debug,
            simplelog::Config::default(),
            simplelog::TerminalMode::Mixed,
            simplelog::ColorChoice::Auto,
        ),
        WriteLogger::new(
            simplelog::LevelFilter::Info,
            simplelog::Config::default(),
            File::create("budget_rs.log").unwrap(),
        ),
    ])
    .unwrap();

    // The web server
    let _ = HttpServer::new(|| {
        App::new()
            .route("/savings_input", web::post().to(savings_post)) // GET endpoint for htmx
            .service(web::resource("/").to(|| async {
                HelloTemplate {
                    title: "Hunter Home",
                    index: titled,
                } // Accessable in the HTML template; askama
            }))
            .service(home_button) // When home button is pressed.
            .service(web::resource("/about").to(|| async {
                About {
                    title: "About",
                    about: abouted,
                } // Accessable in the HTML template; askama
            }))
            .service(web::resource("/debt").to(|| async {
                Debt {
                    title: "Debt",
                    debt: debted,
                } // Accessable in the HTML template; askama
            }))
            .service(web::resource("/savings").to(|| async {
                Savings {
                    title: "Savings",
                    savings: savinged,
                } // Accessable in the HTML template; askama
            }))
            .service(web::resource("/income").to(|| async {
                Income {
                    title: "Income",
                    income: incomed,
                } // Accessable in the HTML template; askama
            }))
            .service(web::resource("/investment").to(|| async {
                Investment {
                    title: "Investment",
                    investment: invested,
                } // Accessable in the HTML template; askama
            }))
            .service(web::resource("/retirement").to(|| async {
                Retirement {
                    title: "Retirement",
                    retirement: retired,
                } // Accessable in the HTML template; askama
            }))
            .service(web::resource("/budget").to(|| async {
                Budget {
                    title: "Budget",
                    budget: budgeted,
                } // Accessable in the HTML template; askama
            }))
            .service(web::resource("/404").to(|| async {
                NotFound {
                    title: "404",
                    not_found: not_founded,
                } // Accessable in the HTML template; askama
            }))
    })
    .workers(1)
    .bind(("127.0.0.1", 8123))
    .unwrap_or_else(|_| {
        log::warn!("Can not bind to port 8123");
        std::process::exit(1);
    })
    .run()
    .await;

    Ok(())
}
